hikari/impl/stateful_guild_chunker.py

```diff
@@ -23,65 +23,298 @@
+        include_presences: undefined.UndefinedOr[bool] = undefined.UNDEFINED,
+        query_limit: int = 0,
+        query: str = "",
+        users: undefined.UndefinedOr[typing.Sequence[snowflakes.SnowflakeishOr[users_.User]]] = undefined.UNDEFINED,
+    ) -> None:
+        super().__init__(app=app, event_type=shard_events.MemberChunkEvent, limit=limit, timeout=timeout)
+        self._guild_id = guild_id
+        self._include_presences = include_presences
+        self._limit = query_limit
+        self._missing_chunks: typing.Optional[typing.MutableSequence[int]] = None
+        self._nonce = date.uuid()
+        self._query = query
+        self._users = users
+        self.filter(lambda event: event.nonce == self._nonce)
+
+    async def _listener(self, event: shard_events.MemberChunkEvent) -> None:
+        if not self._filters(event):
+            return
+
+        if self._missing_chunks is None:
+            self._missing_chunks = list(range(event.chunk_count))
+
+        self._missing_chunks.remove(event.chunk_index)
+
+        try:
+            self._queue.put_nowait(event)
+        except asyncio.QueueFull:
+            pass
+
+        if not self._missing_chunks:
+            # While it is tempting to just close the stream here, that'd lead to a type error being
+            # raised the next time they try to iterate over it.
+            assert self._registered_listener is not None
+            self._app.dispatcher.unsubscribe(shard_events.MemberChunkEvent, self._registered_listener)
+
+    async def __anext__(self) -> shard_events.MemberChunkEvent:
+        if not self._active:
+            raise TypeError("stream must be started with `await with` before entering it")
+
+        if self._queue.empty() and not self._missing_chunks and self._missing_chunks is not None:
+            raise StopAsyncIteration from None
+
+        try:
+            return await asyncio.wait_for(self._queue.get(), timeout=self._timeout)
+        except asyncio.TimeoutError:
+            raise StopAsyncIteration from None
+
+    async def open(self) -> None:
+        # as super().open will possibly flip self.started, we want to have its state when this was originally called.
+        started = self._active
+        await super().open()
+
+        if not started:
+            shard_id = snowflakes.calculate_shard_id(self._app, self._guild_id)
+            await self._app.shards[shard_id].request_guild_members(
+                guild=self._guild_id,
+                include_presences=self._include_presences,
+                query=self._query,
+                limit=self._limit,
+                users=self._users,
+                nonce=self._nonce,
+            )
+
 
-_LOGGER = logging.getLogger("hikari.guild_chunker")
+@attr.s(init=True, kw_only=True, hash=False, eq=True, slots=True, weakref_slot=False)
+class _TrackedRequests:
+    """`hikari.api.chunker.RequestInformation` implementation."""
+
+    __slots__: typing.Sequence[str]
```

Comment:
hmmmmm.

Reply:
If that isn't included __slots__ will be treated as something like `typing.Union[typing.Iterable[str], str]]` which isn't compatible with the protocol which types __slots__ as typing.Sequence[str]